### PR TITLE
[type] Decouple quant from SNode 4/n: Add exponent info to BitStructType

### PR DESCRIPTION
**Pull request by strongoier**
Related issue = #4857

Previously, `BitStructType` only contained member types and member offsets, which were not enough to handle `QuantFloatType`, making quant codegen rely on SNode info. This PR adds exponent info to `BitStructType` so that it becomes possible to handle `QuantFloatType` (with or without shared exponent) without SNode info.

### CODE REVIEW

File: taichi/ir/type.cpp [hunk 1]
```diff
@@ -166,12 +166,19 @@ bool QuantFloatType::get_is_signed() const {
...
+    const std::vector<int> &member_exponents,
+    const std::vector<std::vector<int>> &member_exponent_users)
     : physical_type_(physical_type),
       member_types_(member_types),
-      member_bit_offsets_(member_bit_offsets) {
+      member_bit_offsets_(member_bit_offsets),
+      member_owns_shared_exponents_(member_owns_shared_exponents),
+      member_exponents_(member_exponents),
+      member_exponent_users_(member_exponent_users) {
   TI_ASSERT(member_types_.size() == member_bit_offsets_.size());
```
Comment (ailzhang) [inline taichi/ir/type.cpp:182]: nit: do we need to make check more sizes given that we now have more vectors?

File: tests/cpp/ir/type_test.cpp [hunk 2]
```diff
@@ -18,10 +18,10 @@ TEST(Type, BitTypes) {
   auto qu11 = TypeFactory::get_instance().get_quant_int_type(11, false, i32);
   auto u16 = TypeFactory::get_instance().get_primitive_int_type(16, false);
 
-  auto bs =
-      TypeFactory::get_instance().get_bit_struct_type(u16, {qi5, qu11}, {0, 5});
+  auto bs = TypeFactory::get_instance().get_bit_struct_type(
+      u16, {qi5, qu11}, {0, 5}, {false, false}, {-1, -1}, {{}, {}});
 
-  EXPECT_EQ(bs->to_string(), "bs(qi5@0, qu11@5)");
```
Comment (ailzhang) [inline tests/cpp/ir/type_test.cpp:24]: nit: how about adding an additional check for bit_struct_type with shared exponent?

**Review by ailzhang — APPROVED** [review]
LGTM!

File: taichi/ir/type.cpp (hunk 1 above)
Comment (strongoier) [reply taichi/ir/type.cpp:182]: done

**Pull request merged**